src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -2225,6 +2227,142 @@ class JNILocalRootsClosure : public OopClosure {
   virtual void do_oop(narrowOop* obj_p) { ShouldNotReachHere(); }
 };
 
+// Helper class to collect/report stack roots.
+class StackRootCollector {
+private:
+  JvmtiTagMap* tag_map;
+  JNILocalRootsClosure* blk;
+  // java_thread is needed only to report JNI local on top native frame;
+  // I.e. it's required only for platform/carrier threads or mounted virtual threads.
+  JavaThread* java_thread;
+
+  oop threadObj;
+  jlong thread_tag;
+  jlong tid;
+
+  bool is_top_frame;
+  int depth;
+  frame* last_entry_frame;
+
+public:
+  StackRootCollector(JvmtiTagMap* tag_map, JNILocalRootsClosure* blk, JavaThread* java_thread)
+    : tag_map(tag_map), blk(blk), java_thread(java_thread),
+      threadObj(nullptr), thread_tag(0), tid(0),
+      is_top_frame(true), depth(0), last_entry_frame(nullptr)
+  {
+  }
+
+  bool set_thread(oop o);
+  // sets the thread and reports the reference to it with the specified kind.
+  bool set_thread(jvmtiHeapReferenceKind kind, oop o);
+
+  bool do_frame(vframe* vf);
+};
+
+bool StackRootCollector::set_thread(oop o) {
+  threadObj = o;
+  thread_tag = tag_for(tag_map, threadObj);
+  tid = java_lang_Thread::thread_id(threadObj);
+
+  is_top_frame = true;
+  depth = 0;
+  last_entry_frame = nullptr;
+
+  return true;
+}
+
+bool StackRootCollector::set_thread(jvmtiHeapReferenceKind kind, oop o) {
+  return set_thread(o)
+         && CallbackInvoker::report_simple_root(kind, threadObj);
+}
+
+bool StackRootCollector::do_frame(vframe* vf) {
+  if (vf->is_java_frame()) {
+    // java frame (interpreted, compiled, ...)
+    javaVFrame* jvf = javaVFrame::cast(vf);
+
+    // the jmethodID
+    jmethodID method = jvf->method()->jmethod_id();
+
+    if (!(jvf->method()->is_native())) {
+      jlocation bci = (jlocation)jvf->bci();
+      StackValueCollection* locals = jvf->locals();
+      for (int slot = 0; slot < locals->size(); slot++) {
+        if (locals->at(slot)->type() == T_OBJECT) {
+          oop o = locals->obj_at(slot)();
+          if (o == nullptr) {
+            continue;
+          }
+
+          // stack reference
+          if (!CallbackInvoker::report_stack_ref_root(thread_tag, tid, depth, method,
+                                                      bci, slot, o)) {
+            return false;
+          }
+        }
+      }
+
+      StackValueCollection* exprs = jvf->expressions();
+      for (int index = 0; index < exprs->size(); index++) {
+        if (exprs->at(index)->type() == T_OBJECT) {
+          oop o = exprs->obj_at(index)();
+          if (o == nullptr) {
+            continue;
+          }
+
+          // stack reference
+          if (!CallbackInvoker::report_stack_ref_root(thread_tag, tid, depth, method,
+                                                      bci, locals->size() + index, o)) {
+            return false;
+          }
+        }
+      }
```

Comment:
refactored.

Reply:
It'd be nice to do even more factoring + renaming.
The lines 2326-2345 can be refactored to a function:
```
  bool StackRootCollector::report_native_frame_refs(jmethodID method) {
    _blk->set_context(_thread_tag, _tid, _depth, method);
    if (_is_top_frame) {
      // JNI locals for the top frame.
      assert(_java_thread != nullptr, "sanity");
      _java_thread->active_handles()->oops_do(_blk);
      if (_blk->stopped()) {
        return false;
      }
    } else {
      if (_last_entry_frame != nullptr) {
        // JNI locals for the entry frame
        assert(_last_entry_frame->is_entry_frame(), "checking");
        _last_entry_frame->entry_frame_call_wrapper()->handles()->oops_do(_blk);
        if (_blk->stopped()) {
          return false;
        }
      }
    }
    return true;
  }
```

The function `report_stack_refs` can be renamed to `report_java_frame_refs`
to make function name more consistent.

Reply:
Sorry, I do not see how this improves readability.
Big functions with many layered conditions do not improve readability.

Reply:
I mean the pieces of the code that set and use _is_top_frame/_last_entry_frame are close so it's easier to see the logic

Reply:
I'd say that it will be even better to find out what are manipulations with these instance fields. They are defined in class scope anyway. Also, you can place the definition of function `report_native_frame_refs()` right after `do_frame()` definition, so they occurrences will be still close.
I think, it is more important to see the whole logics of the `do_frame()` with less cascading levels.
You can give it a try and see the advantage.

Reply:
fixed